### PR TITLE
Add FXIOS-8478 [v124] Bottomsheet for address autofill and inject forms upon selection of address

### PR DESCRIPTION
**Pull request by jnrahme**
## :scroll: Tickets
[Jira ticket](https://mozilla-hub.atlassian.net/browse/FXIOS-8478)
[Github issue](https://github.com/mozilla-mobile/firefox-ios/issues/18827)

## :bulb: Description
As an application engineer, I want to add a feature where users can easily input their address. So, when they tap on a button or a specific area in the app, a little window at the bottom of the screen will pop up. This window will show a list of addresses to choose from. Once they pick their address from this list, the form on the screen will automatically fill in all the necessary details like the street name, city, country, and so on. This way, users don't have to type everything manually, making it quicker and more convenient for them to enter their address.
Screenshot 2024-02-23 at 2 22 59 PM

## :pencil: Checklist
You have to check all boxes before merging
- [x] Filled in the above information (tickets numbers and description of your work)
- [x] Updated the PR name to follow our [PR naming guidelines](https://github.com/mozilla-mobile/firefox-ios/wiki/Pull-Request-Naming-Guide)
- [ ] Wrote unit tests and/or ensured the tests suite is passing
- [ ] When working on UI, I checked and implemented accessibility (minimum Dynamic Text and VoiceOver)
- [ ] If needed, I updated documentation / comments for complex code and public methods
- [ ] If needed, added a backport comment (example `@Mergifyio backport release/v120`)

### CODE REVIEW

**Comment by jnrahme** [comment]
https://github.com/mozilla-mobile/firefox-ios/assets/11262696/0698638b-3873-4fd7-8997-9a5cd08ee8a1

File: firefox-ios/Client/Coordinators/AddressAutofillCoordinator.swift [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+        self.themeManager = themeManager
+        self.tabManager = tabManager
+        self.parentCoordinator = parentCoordinator
+        super.init(router: router)
+    }
+
+    // MARK: - Methods
+
+    /// Shows the address autofill functionality within a bottom sheet.
+    /// - Parameter frame: The WKFrameInfo object representing the web view frame.
```
Comment (nbhasin2) [inline firefox-ios/Client/Coordinators/AddressAutofillCoordinator.swift:46]: I believe we talked about this, you don't really need comments for functions if they are self explanatory.
this one seems pretty straight forward no? 🤔 @jnrahme

File: firefox-ios/Client/Coordinators/AddressAutofillCoordinator.swift [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+    func showAddressAutofill(frame: WKFrameInfo?) {
+        let bottomSheetViewModel = BottomSheetViewModel(closeButtonA11yLabel: .CloseButtonTitle)
+//        bottomSheetViewModel.shouldDismissForTapOutside = false
```
Comment (nbhasin2) [inline firefox-ios/Client/Coordinators/AddressAutofillCoordinator.swift:49]:
```suggestion
```
do you need this commented out part?

File: firefox-ios/Client/Coordinators/AddressAutofillCoordinator.swift [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+
+        let viewModel = AddressListViewModel(profile: profile) // Initialize your view model
+
+        // Set the address selection callback closure
+        viewModel.addressSelectionCallback = { [weak self] selectedAddress in
```
Comment (lmarceau) [inline firefox-ios/Client/Coordinators/AddressAutofillCoordinator.swift:54]: nit; I would argue those comments are not needed and self explanatory as well
```suggestion
        let viewModel = AddressListViewModel(profile: profile)

        viewModel.addressSelectionCallback = { [weak self] selectedAddress in
```

File: firefox-ios/Client/Coordinators/AddressAutofillCoordinator.swift [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+            FormAutofillHelper.injectAddressInfo(logger: self.logger,
+                                                 address: selectedAddress,
+                                                 tab: currentTab,
+                                                 frame: frame) { error in
+                guard let error = error else {
+                    return
+                }
+                self.logger.log("Address bottom sheet injection \(error)",
+                                level: .debug,
+                                category: .webview)
```
Comment (lmarceau) [inline firefox-ios/Client/Coordinators/AddressAutofillCoordinator.swift:71]: I don't think the logger category should be `.webview` here. I think the `LoggerCategory` `creditcard` could be renamed to `autofill` to encompass credit card and address autofill. Would be better suited than webview for sure.
```suggestion
                                category: .webview)
```

File: firefox-ios/Client/Frontend/TabContentsScripts/FormAutofillHelper/FormAutofillHelper.swift [hunk 5]
```diff
@@ -165,8 +165,41 @@ class FormAutofillHelper: TabContentScript {
...
+            webView.evaluateJavascriptInDefaultContentWorld(fillAddressInfoCallback, frame) { _, error in
+                if let error = error {
+                    completion(error)
+                    logger.log("Address script error \(error)", level: .debug, category: .webview)
+                } else {
+                    completion(nil)
+                }
+            }
+        } catch let error as NSError {
+            logger.log("Address script error \(error)", level: .debug, category: .webview)
```
Comment (lmarceau) [inline firefox-ios/Client/Frontend/TabContentsScripts/FormAutofillHelper/FormAutofillHelper.swift:199]: Same comment for those logger call, category should be adjusted.

File: firefox-ios/Client/Frontend/TabContentsScripts/FormAutofillHelper/FormAutofillHelper.swift [hunk 6]
```diff
@@ -165,8 +165,41 @@ class FormAutofillHelper: TabContentScript {
 
         return AutofillFieldValuePayload(fieldValue: .address, fieldData: addressPlainText)
     }
+
     // MARK: - Injection
 
+    static func injectAddressInfo(logger: Logger,
```
Comment (lmarceau) [inline firefox-ios/Client/Frontend/TabContentsScripts/FormAutofillHelper/FormAutofillHelper.swift:171]: Why static? Can we avoid it?

File: firefox-ios/Client/Frontend/Autofill/Address/AddressListViewModel.swift [hunk 7]
```diff
@@ -62,4 +63,27 @@ class AddressListViewModel: ObservableObject {
...
+    }
+
+    // MARK: - Handle Address Selection
+
+        /// Handles the selection of an address.
+        /// - Parameter address: The selected address.
+        func handleAddressSelection(_ address: Address) {
+            // Perform actions with the selected address
+            addressSelectionCallback?(fromAddress(address))
+        }
```
Comment (lmarceau) [inline firefox-ios/Client/Frontend/Autofill/Address/AddressListViewModel.swift:88]: indentation + comment that isn't needed IMO
```suggestion
        /// Handles the selection of an address.
        /// - Parameter address: The selected address.
        func handleAddressSelection(_ address: Address) {
            addressSelectionCallback?(fromAddress(address))
        }
```

File: firefox-ios/Client/Coordinators/Browser/BrowserCoordinator.swift [hunk 8]
```diff
@@ -562,11 +562,32 @@ class BrowserCoordinator: BaseCoordinator,
...
         bottomSheetCoordinator.showPassCodeController()
     }
 
+    private func makeAddressAutofillCoordinator() -> AddressAutofillCoordinator {
+        if let bottomSheetCoordinator = childCoordinators.first(where: {
+            $0 is AddressAutofillCoordinator
+        }) as? AddressAutofillCoordinator {
+            return bottomSheetCoordinator
+        }
+        let bottomSheetCoordinator = AddressAutofillCoordinator(
```
Comment (lmarceau) [inline firefox-ios/Client/Coordinators/Browser/BrowserCoordinator.swift:581]: Just to make sure, we need to be careful when we spin off coordinators, they need to be properly `deinit` whenever the VC is removed from the view hierarchy (i.e. when the VC is dismissed, the corresponding coordinator should be dismissed as well). There are some exceptions in our codebase where _some_ coordinators are kept in memory once created, but they are an exception to alleviate some previous technical debt. 

I would expect unit tests for any new coordinator code added in `BrowserCoordinator` and new coordinators like the `AddressAutoFillCoordinator`.

File: firefox-ios/Client/Coordinators/AddressAutofillCoordinator.swift [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+import Common
+import Storage
+import ComponentLibrary
+import WebKit
+
+/// Coordinator responsible for managing the address autofill functionality, presenting it within a bottom sheet.
+class AddressAutofillCoordinator: BaseCoordinator {
+    // MARK: - Properties
+
+    typealias BottomSheetCardParentCoordinator = BrowserNavigationHandler & ParentCoordinatorDelegate
```
Comment (lmarceau) [inline firefox-ios/Client/Coordinators/AddressAutofillCoordinator.swift:14]: I don't think this delegate needs to be a `BrowserNavigationHandler` as well here. Only a `ParentCoordinatorDelegate` seems to suffice?

**Review by lmarceau — CHANGES_REQUESTED** [review]

File: firefox-ios/Client/Frontend/TabContentsScripts/FormAutofillHelper/FormAutofillHelper.swift (hunk 6 above)
Comment (nbhasin2) [reply firefox-ios/Client/Frontend/TabContentsScripts/FormAutofillHelper/FormAutofillHelper.swift:171]: @lmarceau this is one area where avoiding static is not trivial but due to nature of how autofill works, the injection can take place from practically anywhere. Here in this case the flow is `address input form tap -> show accessory view -> bottom sheet -> tap on an address`

Due to the nature of this flow by the time tapping on address happens the only way to trigger an injection is from within that Tab. Which seems feasible however adding more scripts makes it more convoluted inside the tab. 

So even though static might not be the right variable the argument is to put it in one place where it belongs. Here it `best` belongs inside FormAutofillHelper.swift compared to Tab.swift

I am open for suggestion :)

Comment (nbhasin2) [reply firefox-ios/Client/Frontend/TabContentsScripts/FormAutofillHelper/FormAutofillHelper.swift:171]: @jnrahme On a separate note, did we add tests for this? 
Sorry I might have asked before but are you are waiting for any js fixes for this area?

File: firefox-ios/Client/Coordinators/AddressAutofillCoordinator.swift (hunk 1 above)
Comment (jnrahme) [reply firefox-ios/Client/Coordinators/AddressAutofillCoordinator.swift:46]: Yes  but this is a stand-alone class, its not a function within the browserVC

File: firefox-ios/Client/Frontend/TabContentsScripts/FormAutofillHelper/FormAutofillHelper.swift (hunk 6 above)
Comment (jnrahme) [reply firefox-ios/Client/Frontend/TabContentsScripts/FormAutofillHelper/FormAutofillHelper.swift:171]: Added tests

File: firefox-ios/Client/Coordinators/Browser/BrowserCoordinator.swift (hunk 8 above)
Comment (jnrahme) [reply firefox-ios/Client/Coordinators/Browser/BrowserCoordinator.swift:581]: Added test

**Review by lmarceau — APPROVED** [review]

**Review by nbhasin2 — APPROVED** [review]

**Comment by clarmso** [comment]
I'm trying to resolve some Bitrise-specific build issues. I could compile this branch on my computer, but Bitrise repeatedly gives compilation errors.

**Comment by nbhasin2** [comment]
@mergify backport release/v124

**Comment by jnrahme** [comment]
@mergify backport release/v124

**Pull request merged**